Wasm threads blog post (#5693)
GoogleChrome/web.dev: src/site/content/en/blog/webassembly-threads/index.md

---

> title: Using WebAssembly threads from C, C++ and Rust

Reviewer: ```suggestion
title: Using WebAssembly threads from C, C++ and Rust
subhead: Learn how to bring multithreaded applications written in other languages to WebAssembly.
```

Author: Should it really duplicate the description? Is one derived from the other?

Reviewer: Description is for search engines, should be 1-3 sentence summary of the post. The subhead is what's displayed on the blog page, so it's a chance to give a bit more info than what fits in the title. They can be identical, though description could be more descriptive too. :)

description: Learn how to bring multithreaded applications written in other languages to WebAssembly.
date: 2021-06-22
hero: image/9oK23mr86lhFOwKaoYZ4EySNFp02/YrOqDnzjHFqmZdiNBmbw.jpg
alt: A needle acting as a prism—splitting a single white thread into multiple colourful ones.
authors:
  - rreverser
tags:
  - blog # blog is a required tag for the article to show up in the blog.
  - webassembly
  - performance
---

WebAssembly threads support is one of the most important performance additions to WebAssembly. It
allows you to either run parts of your code in parallel on separate cores, or the same code over
independent parts of the input data, scaling it to as many cores as the user has and significantly
reducing the overall execution time.

In this article you will learn how to use WebAssembly threads to bring multithreaded applications

> written in languages like C, C++ and Rust to the Web.

Reviewer: @rachelandrew The "Web" shouldn't be capitalized, refer to the word list https://web.dev/handbook/word-list/ and avoid unnecessary capitalization in general.

## How WebAssembly threads work

WebAssembly threads is not a separate feature, but a combination of several components that allows

> WebAssembly apps to use traditional multithreading paradigms on the Web.

Reviewer: ```suggestion
WebAssembly apps to use traditional multithreading paradigms on the web.
```

### Web Workers

First component is the regular
[Workers](https://developer.mozilla.org/docs/Web/API/Web_Workers_API/Using_web_workers) you
know and love from JavaScript. WebAssembly threads use the `new Worker` constructor to create new
underlying threads. Each thread loads a JavaScript glue, and then the main thread uses
[`Worker#postMessage`](https://developer.mozilla.org/docs/Web/API/Worker/postMessage) method
to share the compiled
[`WebAssembly.Module`](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/WebAssembly/Module)
as well as a shared
[`WebAssembly.Memory`](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/WebAssembly/Memory)
(see below) with those other threads. This establishes communication and allows all those threads to
run the same WebAssembly code on the same shared memory without going through JavaScript again.

Web Workers have been around for over a decade now, are widely supported, and don't require any
special flags.

> ### SharedArrayBuffer

Reviewer: ```suggestion
### `SharedArrayBuffer`
```

Fast-forward to 2020, Chrome and Firefox both have implementations of Site Isolation now and also
decided on a standard way for websites to opt-in to the feature: [COOP/COEP
headers](https://web.dev/coop-coep/). An opt-in mechanism allows to use Site Isolation even on
low-powered devices where enabling it for all the websites would be too expensive. To opt-in, add
the following headers to the main document in your server configuration:

```http
Cross-Origin-Embedder-Policy: require-corp
Cross-Origin-Opener-Policy: same-origin
```

Once you opt-in, you get access to `SharedArrayBuffer` (including `WebAssembly.Memory` backed by a
`SharedArrayBuffer`), precise timers, memory measurement and other APIs that require an isolated
origin for security reasons. Check out the [linked guide](https://web.dev/coop-coep/) for more
details.

### WebAssembly atomics

While `SharedArrayBuffer` allows each thread to read and write to the same memory, for correct
communication you want to make sure they don't "step on each others' toes". For example, it's
possible for one thread to start reading data from a shared address, while another thread is writing

> to it, so the first thread will now get a corrupted result. This category of bugs is known as race

Author: This rewording changes a meaning, we definitely want threads acting on the same thing, but at the same time we want to make sure that whatever one thread is doing doesn't conflict with what another is doing while they are both touching the same thing.

> to it, so the first thread will now get a corrupted result. This category of bugs is known as race

Author: I'm going to close this one.

> to it, so the first thread will now get a corrupted result. This category of bugs is known as race

Reviewer 2: What I'm trying to avoid is the phrase "step on each other's toes" do you have a suggestion for something clearer?

> to it, so the first thread will now get a corrupted result. This category of bugs is known as race

Reviewer 2: @RReverser I'd like to find an alternative to "step on each other's toes" it is a colloquialism and something of an anthropomorphism (https://developers.google.com/style/anthropomorphism) given that threads don't have toes :) . This is what I was trying to get at, do you have a more descriptive way of saying this?

> While `SharedArrayBuffer` allows each thread to read and write to the same memory, for correct
> communication you want to make sure they don't "step on each others' toes". For example, it's
> possible for one thread to start reading data from a shared address, while another thread is writing
> to it, so the first thread will now get a corrupted result. This category of bugs is known as race

Author: Ah. I Love adding some expressions like this, it feels more... human I guess. But you're right, it might not be widely understood and goes against the style. I can do this:

```suggestion
While `SharedArrayBuffer` allows each thread to read and write to the same memory, for correct
communication you want to make sure they don't perform conflicting operations at the same time. For example, it's
possible for one thread to start reading data from a shared address, while another thread is writing
to it, so the first thread will now get a corrupted result. This category of bugs is known as race
```

conditions. In order to prevent race conditions, you need to somehow synchronize those accesses.
This is where atomic operations come in.

[WebAssembly
atomics](https://webassembly.github.io/threads/core/syntax/instructions.html#atomic-memory-instructions)
is an extension to the WebAssembly instruction set that allow to read and write small cells of data
(usually 32- and 64-bit integers) "atomically", that is, in a way that guarantees that no two
threads are reading / writing to the same cell at the same time, preventing such conflicts at a low

> level. Additionally, WebAssembly atomics contain two more instruction kinds—"wait" and "notify"—that

Author: I thought types might be a bit more ambigous, because it's not really new high-level type of instructions (like we have memory instructions, control flow instructions etc.). I guess it's ambigous either way and doesn't matter too much, but "kind" seems a bit more lightweight.

To ensure that all users can load your application, you'll need to implement progressive enhancement
by building two different versions of Wasm—one with multithreading support and one without it. Then
load the supported version depending on feature detection results. To detect WebAssembly threads
support at runtime, use [wasm-feature-detect
library](https://github.com/GoogleChromeLabs/wasm-feature-detect) and load the module like this:

```js
import { threads } from 'wasm-feature-detect';

const hasThreads = await threads();

const module = await (
  hasThreads
    ? import('./module-with-threads.js')
    : import('./module-without-threads.js')
);

// …now use `module` as you normally would
```

Now let's take a look at how to build a multithreaded version of the WebAssembly module.

### C

In C, particularly on Unix-like systems, the common way to use threads is via [POSIX
Threads](https://en.wikipedia.org/wiki/POSIX_Threads) provided by the `pthread` library. Emscripten
[provides an API-compatible implementation](https://emscripten.org/docs/porting/pthreads.html) of
the `pthread` library built atop Web Workers, shared memory and atomics, so that the same code can

> work on the Web without changes.

Reviewer: ```suggestion
work on the web without changes.
```

Let's take a look at an example:

> 

Author: TIL, nice.

```c/2,17,19
#include <stdio.h>
#include <unistd.h>
#include <pthread.h>

void *thread_callback(void *arg)
{
    sleep(1);
    printf("Inside the thread: %d\n", *(int *)arg);
    return NULL;
}

int main()
{
    puts("Before the thread");

    pthread_t thread_id;
    int arg = 42;
    pthread_create(&thread_id, NULL, thread_callback, &arg);

    pthread_join(thread_id, NULL);

    puts("After the thread");

    return 0;
}
```

> Here the headers for the pthread library are included via `pthread.h`. You can also see a couple of
> crucial functions for dealing with threads.

Reviewer: ```suggestion
Here the headers for the `pthread` library are included via `pthread.h`. You can also see a couple of
crucial functions for dealing with threads.
```

[`pthread_create`](https://man7.org/linux/man-pages/man3/pthread_create.3.html) will create a
background thread. It takes a destination to store a thread handle in, some thread creation
attributes (here not passing any, so it's just `NULL`), the callback to be executed in the new
thread (here `thread_callback`), and an optional argument pointer to pass to that callback in case
you want to share some data from the main thread - in this example we're sharing a pointer to a

> variable `arg`.

Reviewer: + 1, it's difficult to follow as is.

[`pthread_join`](https://man7.org/linux/man-pages/man3/pthread_join.3.html) can be called later at
any time to wait for the thread to finish the execution, and get the result returned from the
callback. It accepts the earlier assigned thread handle as well as a pointer to store the result in

> - in this case I don't have any results, so I'm just passing `NULL`.

Reviewer: This is not a rule we enforce strictly, but try to avoid speaking in first person in code instructions (both "we" and "I"). For example, you could rephrase like this: "In this case, there aren't any results so the function takes a `NULL` argument".

> - in this case I don't have any results, so I'm just passing `NULL`.

Author: Good point, thanks.

> 

Reviewer: Consider adding another heading level here, for example "Compiling C code with Emscripten".

Author: Both C and C++ sections are all about Emscripten, so I don't think another heading is necessary here.

To compile code using threads with Emscripten, you need to invoke `emcc` and pass a `-pthread`
parameter like you would when compiling same code with Clang or GCC on other platforms:

> ```shell

Reviewer: Use some of the formatting options to emphasise that this is not the best way to do it. https://web.dev/handbook/web-dev-components/#asides

Author: As mentioned below, I've just removed asides for now as I can't come up with a good non-distracting usage in this post.

emcc -pthread example.c -o example.js
```

However, when you try to run it in a browser or Node.js, you'll see a warning and then the program
will hang:

```text
Before the thread
Tried to spawn a new thread, but the thread pool is exhausted.
This might result in a deadlock unless some threads eventually exit or the code
explicitly breaks out to the event loop.
If you want to increase the pool size, use setting `-s PTHREAD_POOL_SIZE=...`.
If you want to throw an explicit error instead of the risk of deadlocking in those
cases, use setting `-s PTHREAD_POOL_SIZE_STRICT=2`.
[…hangs here…]
```

> What happened? The problem is, most of the time-consuming APIs on the Web are asynchronous and rely

Reviewer: ```suggestion
What happened? The problem is, most of the time-consuming APIs on the web are asynchronous and rely
```

One way to solve this problem is to create a pool of Workers ahead of time, before the program has
even started, so that by the time `pthread_create` is invoked, it only needs to take a ready-to-use
Worker from the pool, run the provided callback on that background thread and return the Worker back
to the pool. All of this can be done synchronously, so there won't be any deadlocks as long as the
pool is sufficiently large.

This is exactly what Emscripten allows with the [`-s
PTHREAD_POOL_SIZE=...`](https://emsettings.surma.technology/#PTHREAD_POOL_SIZE) option. It allows to
specify a number of threads - either a fixed number, or a JavaScript expression like
[`navigator.hardwareConcurrency`](https://developer.mozilla.org/docs/Web/API/NavigatorConcurrentHardware/hardwareConcurrency)
to create as many threads as there are cores on the CPU. The latter option is helpful when your code
can scale to an arbitrary number of threads.

In the example above, there is only one thread being created, so instead of reserving all cores it's
sufficient to use `-s PTHREAD_POOL_SIZE=1`:

```shell
emcc -pthread -s PTHREAD_POOL_SIZE=1 example.c -o example.js
```

> {% Aside 'success' %} This time, when you execute it, things work successfully: {% endAside %}

Reviewer: Could you make this aside more descriptive? Something along the lines "When you execute [action] in a [correct way to do it] a [good thing happens (or a bad thing is avoided)]."

I'd also suggest to use "caution" or "gotcha" formatting around the previous example to contrast them.

```text
Before the thread
Inside the thread: 42
After the thread
Pthread 0x701510 exited.
```

{% Aside 'caution' %} There is another problem though: see that `sleep(1)` in the code example? It
executes in the thread callback, meaning off the main thread, so it should be fine, right? Well, it

> isn't. {% endAside %}

Reviewer: Agreed. Another thing that's not great is that the aside is referring to a line in a code example that's a couple of scrolls above. It's a lot to expect from readers to keep in working memory while reading and it's potentially disorienting to scroll back to refer to that place in code. 

Suggestion: What Rachel said + repeat that function `void *thread_callback` and highlight the relevant line. Consider adding another subheading level for that.

> isn't. {% endAside %}

Author: I agree, I also mentioned in the email that styling of asides didn't feel quite right here but thought maybe it's worth using it to highlight the problematic example anyway. I wonder if it's easier to just remove the aside again like it was in the draft...

> isn't. {% endAside %}

Author: I went forth and back on those, and while some highlighting seemed useful, I just couldn't come up with any usage of Asides here that actually helps rather than distracting from the flow, so on the 2nd thought just removed them again.

When `pthread_join` is called, it has to wait for the thread execution to finish, meaning that if
the created thread is performing long-running tasks—in this case, sleeping 1 second—then the main
thread will also have to block for the same amount of time till the results are back. When this JS

> is executed in the browser, it will block the UI thread for 1 second until the thread callback

Author: Should the "1 second" in the previous sentence be changed as well for symmetry?

> is executed in the browser, it will block the UI thread for 1 second until the thread callback

Reviewer: Actually, because this is a measurement it should be "1 second" per Google styleguide. @rachelandrew Here's the reference: https://developers.google.com/style/numbers?hl=en#numbers-as-numerals

returns. This leads to poor user experience.

> There are few solutions to this.

Reviewer: Can you briefly introduce the solutions in a list? It can be as simple as:
- `pthread_detach`
- `-s PROXY_TO_PTHREAD`
- create your own Worker and expose it with Comlink

Author: I like this.

Author: Hm, need to think about how to link back to corresponding paragraphs though. Should those be subheadings?

Reviewer: Yeah, just make sure those subheading have the correct level.

Author: Done

`pthread` library under the hood.

So the example above can be rewritten in more idiomatic C++ like this:

```cpp/1,9,14
#include <iostream>
#include <thread>
#include <chrono>

int main()
{
    puts("Before the thread");

    int arg = 42;
    std::thread thread([&]() {
        std::this_thread::sleep_for(std::chrono::seconds(1));
        std::cout << "Inside the thread: " << arg << std::endl;
    });

    thread.join();

    std::cout << "After the thread" << std::endl;

    return 0;
}
```

When compiled and executed with similar parameters, it will behave in the same way as the C example:

```shell
emcc -std=c++11 -pthread -s PROXY_TO_PTHREAD example.cpp -o example.js
```

Output:

```text
Before the thread
Inside the thread: 42
Pthread 0xc06190 exited.
After the thread
Proxied main thread 0xa05c18 finished with return code 0. EXIT_RUNTIME=0 set, so keeping
main thread alive for asynchronous event operations.
Pthread 0xa05c18 exited.
```

### Rust

Unlike Emscripten, Rust doesn't have a specialised end-to-end Web target, but instead provides a
generic `wasm32-unknown-unknown` target for generic WebAssembly output.

> If Wasm is intended to be used in a Web environment, any interaction with JavaScript APIs is left to

Reviewer: ```suggestion
If Wasm is intended to be used in a web environment, any interaction with JavaScript APIs is left to
```

Author: Looks like this comment was left on outdated text, it already has "web" lowercase in latest version.

external libraries and tooling like [wasm-bindgen](https://rustwasm.github.io/docs/wasm-bindgen/)
and [wasm-pack](https://rustwasm.github.io/docs/wasm-pack/). Unfortunately, this means that the
standard library is not aware of Web Workers and standard APIs like
[`std::thread`](https://doc.rust-lang.org/std/thread/) won't work when compiled to WebAssembly.

Luckily, the majority of the ecosystem depends on higher-level libraries to take care of
multithreading. At that level it's much easier to abstract away all the platform differences.

In particular, [Rayon](https://crates.io/crates/rayon) is the most popular choice for
data-parallelism in Rust. It allows to take method chains on regular iterators and, usually with a
single line change, convert them in a way where they'd run in parallel on all available threads
instead of sequentially. For example:

```rust/3/2
pub fn sum_of_squares(numbers: &[i32]) -> i32 {
  numbers
  .iter()
  .par_iter()
  .map(x => x * x)
  .sum()
}
```

> This unassuming code will split up the input data, calculate `x * x` and partial sums in parallel
> threads, and in the end add up those partial results together.

Reviewer: ```suggestion
This code will split up the input data, calculate `x * x` and partial sums in parallel
threads, and in the end add up those partial results together.
```

Reviewer: Unnecessary adjective.

Author: I somehow want to emphasise that the difference is really just a `par_` prefix that makes all of this to work, even though we're not manually creating and dealing with threads anywhere like we did in C/C++ examples.

Maybe rewording to "With this small change, the code ..." would be better?

Reviewer: If the emphasis is important, then yes, that's better phrasing. Go for it 👍

> In order to accomodate for platforms without working `std::thread`, Rayon provides hooks that allow

Reviewer: ```suggestion
To accommodate for platforms without working `std::thread`, Rayon provides hooks that allow
```

> to define custom logic for spawning and exiting threads. We've tapped into those hooks and built an

Reviewer: It's okay to use "we" here, but you need to define who "we" are.

> to define custom logic for spawning and exiting threads. We've tapped into those hooks and built an

Author: Hm, in this case technically it's just me. Should I use "I" instead or attribute it to Chrome or Chrome DevRel or ..?

> to define custom logic for spawning and exiting threads. We've tapped into those hooks and built an

Author: Ping - any ideas?

> to define custom logic for spawning and exiting threads. We've tapped into those hooks and built an

Reviewer: If you think the attribution is important in this case, you can do write something like "To [make the developer experience better] I've built the wasm-bindgen-rayon adapter which [does something]. 

If not, you can rewrite the sentence to omit attribution. Something like "You can access those hooks from wasm-bindgen-rayon adapter..."

To use it, you need to re-export an `initThreadPool` function from the wasm-bindgen-rayon library in
your main code and invoke it right after initialization.

```js
import init, { initThreadPool /* ... */ } from './pkg/index.js';

// Regular wasm-bindgen initialization.
await init();

// Thread pool initialization with the given number of threads
// (pass `navigator.hardwareConcurrency` if you want to use all cores).
await initThreadPool(navigator.hardwareConcurrency);

// ...now you can invoke any exported functions as you normally would
```

It will create a pool of Workers and reuse them throughout the lifetime of the program. This
initialization is similar to the `-s PTHREAD_POOL_SIZE=...` option in Emscripten explained earlier,
and needs to happen before the main code to avoid the same deadlocks.

There are few other configurations to set up before you can use multithreading, so make sure to
check the [detailed documentation](https://github.com/GoogleChromeLabs/wasm-bindgen-rayon#usage) for
complete setup steps.

{% Aside 'caution' %} The same
[caveats](https://github.com/GoogleChromeLabs/wasm-bindgen-rayon#caveats) about blocking the main

> thread apply here too. {% endAside %}

Reviewer 2: This is another caution that leads into the text, the same comment as above applies.

for an end-to-end demo showing:

*   [Feature detection of
    threads.](https://github.com/GoogleChromeLabs/wasm-bindgen-rayon/blob/e13485d6d64a062b890f5bb3a842b1fe609eb3c1/demo/wasm-worker.js#L27)
*   [Building single- and multi-threaded versions of the same Rust
    app.](https://github.com/GoogleChromeLabs/wasm-bindgen-rayon/blob/e13485d6d64a062b890f5bb3a842b1fe609eb3c1/demo/package.json#L4-L5)
*   [Loading the JS+Wasm generated by wasm-bindgen in a
    Worker.](https://github.com/GoogleChromeLabs/wasm-bindgen-rayon/blob/e13485d6d64a062b890f5bb3a842b1fe609eb3c1/demo/wasm-worker.js#L28-L31)
*   [Using wasm-bindgen-rayon to initialize a thread
    pool.](https://github.com/GoogleChromeLabs/wasm-bindgen-rayon/blob/e13485d6d64a062b890f5bb3a842b1fe609eb3c1/demo/wasm-worker.js#L32)
*   Using Comlink to [expose Worker's
    API](https://github.com/GoogleChromeLabs/wasm-bindgen-rayon/blob/e13485d6d64a062b890f5bb3a842b1fe609eb3c1/demo/wasm-worker.js#L44-L46)
    to [the main
    thread](https://github.com/GoogleChromeLabs/wasm-bindgen-rayon/blob/e13485d6d64a062b890f5bb3a842b1fe609eb3c1/demo/index.js#L25-L29).

> ## Real-world use-cases

Reviewer: This is the interesting part, I suggest either adding a brief summary of exciting real world use cases to the intro or just moving the entire section at the top of the post. Always try to explain "why" before "how to".

Author: I explained "why" in the very beginning of the post, but specific examples feels like a "nice to have" part rather than something to lead with...

Reviewer: Chatted with Rachel about this, you have a point, disregard the suggestion.

We actively use WebAssembly threads in [Squoosh.app](https://squoosh.app/) for client-side image
compression - in particular, for formats like AVIF (C++), JPEG-XL (C++), OxiPNG (Rust) and WebP v2
(C++). Thanks to multithreading alone, we've seen consistent 1.5x-3x speed-ups (exact ratio differs
per codec), and were able to push those numbers even further by combining WebAssembly threads with
[WebAssembly SIMD](https://v8.dev/features/simd)!

> Google Earth is another notable service that's using WebAssembly threads for its [Web

Reviewer: ```suggestion
Google Earth is another notable service that's using WebAssembly threads for its [web
```

version](https://earth.google.com/web/).

[FFMPEG.WASM](https://ffmpegwasm.github.io/) is a WebAssembly version of a popular
[FFmpeg](https://www.ffmpeg.org/) multimedia toolchain that uses WebAssembly threads to efficiently
encode videos directly in the browser.

There are many more exciting examples using WebAssembly threads out there. Be sure to check out

> their demos and bring your own multithreaded applications and libraries to the Web!

Reviewer: ```suggestion
their demos and bring your own multithreaded applications and libraries to the web!
```

---
title: Using WebAssembly threads from C, C++ and Rust
subhead: Learn how to bring multithreaded applications written in other languages to WebAssembly.
description: Learn how to bring multithreaded applications written in other languages to WebAssembly.

> date: 2021-06-22

Reviewer 2: ```suggestion
date: 2021-07-08
```
